### PR TITLE
Update amp_check_finite_and_scale_op and add an updating_loss_scaling op for static graph amp training.

### PR DESCRIPTION
**Pull request by wzzju**
### PR types

Performance optimization

### PR changes

OPs
### Describe

* Use the `check_finite_and_unscale` op to check infinite grads then unscale grads. 
* Use the `update_loss_scaling` op to update the loss scaling value.
The performance of ResNet50 on Tesla V100-16GB single card:

| Original | Optimized(After this pr) |
| -------------------- | ---------------- |
|      952 images/sec       | 985 images/sec    |

image

After training 120 passes on ResNet50 model, the top-1 accuracy and top-5 accuracy is described as below.
|FP32 or AMP|the num of epoch| top-1 accuracy| top-5 accuracy|
|:---|:---|:---|:---|
|FP32|120|75.674%|92.674%|
|AMP|120|76.067%|92.903%|

* Training loss consistency verification during the whole training process.
image
* Average loss and test accuracy consistency verification after each epoch.
image
image
image
image

### CODE REVIEW

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu [hunk 1]
```diff
@@ -58,8 +69,8 @@ class AmpCheckFiniteAndScaleKernel<platform::CUDADeviceContext, T>
       int block = 512;
```
Comment (Xreki) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cu]: block改成1024试试。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu [hunk 2]
```diff
@@ -58,8 +69,8 @@ class AmpCheckFiniteAndScaleKernel<platform::CUDADeviceContext, T>
...
       int grid = (num + block - 1) / block;
       VLOG(3) << "launch kernel";
-      AmpCheckFiniteAndScale<T><<<grid, block, 0, dev_ctx.stream()>>>(
-          x_data, scale_data, num, found_inf_data, out_data);
+      CheckFiniteAndUnscale<T><<<grid, block, 0, dev_ctx.stream()>>>(
```
Comment (Xreki) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:70]: 可以尝试一定程度的unroll，比如一次更新2个、4个等。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+              "check_finite_and_unscale operator.")
+        .AsDuplicable();
+    AddOutput("FoundInfinite",
+              "(Tensor) 1-dim tensor, contains a bool scalar, which indicates "
+              "if there there is infinite or nan item in input X.");
+    AddComment(R"DOC(
+check_finite_and_unscale operator.
+Check if input X contains all finite data, if yes, scale it by input Scale.
+
+$$Out = X * scale$$
```
Comment (zhiqiu) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cc]: Here should be `x / scale` since the name is `check_finite_and_unscale`

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu (hunk 2 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:70]: 感谢，后续可尝试一下。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu (hunk 1 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cu]: Done.但是好像没什么效果。

File: python/paddle/fluid/contrib/mixed_precision/amp_nn.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    Update loss scaling according to overall gradients. If all gradients is 
+    finite after incr_every_n_steps, loss scaling will increase by incr_ratio. 
+    Otherwise, loss scaling will decrease by decr_ratio after
+    decr_every_n_nan_or_inf steps and each step some gradients are infinite.
+
+    Args:
+        found_inf (Variable): A boolean variable indicates whether 
+                                     there is any infinite gradient.
+        prev_loss_scaling (Variable): Previous loss scaling.
+        num_good_steps (Variable): A variable accumulates good steps in which
```
Comment (zhiqiu) [inline python/paddle/fluid/contrib/mixed_precision/amp_nn.py:73]: Not important. 
Maybe `succeed`-`fail`  is better than `good`-`bad` pair, which means a minimize step succeeded or not. The same in c++ op.

File: paddle/fluid/operators/amp/update_loss_scaling_op.cc [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace ops = paddle::operators;
+
+REGISTER_OPERATOR(
+    update_loss_scaling, ops::UpdateLossScalingOp,
+    ops::UpdateLossScalingOpMaker,
+    paddle::framework::EmptyGradOpMaker<paddle::framework::OpDesc>,
+    paddle::framework::EmptyGradOpMaker<paddle::imperative::OpBase>);
+
+REGISTER_OP_CPU_KERNEL(update_loss_scaling, ops::UpdateLossScalingKernel<float>,
```
Comment (zhiqiu) [inline paddle/fluid/operators/amp/update_loss_scaling_op.cc]: Should we register GPU kernel for `update_loss_scaling` ?

File: paddle/fluid/operators/amp/update_loss_scaling_op.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* found_inf = ctx.Input<Tensor>("FoundInfinite");
+    auto* pre_loss_scaling = ctx.Input<Tensor>("PrevLossScaling");
+    auto* good_in = ctx.Input<Tensor>("InGoodSteps");
+    auto* bad_in = ctx.Input<Tensor>("InBadSteps");
+    auto* updated_loss_scaling = ctx.Output<Tensor>("LossScaling");
+    auto* good_out = ctx.Output<Tensor>("OutGoodSteps");
+    auto* bad_out = ctx.Output<Tensor>("OutBadSteps");
+
+    const bool* found_inf_v = found_inf->data<bool>();
```
Comment (zhiqiu) [inline paddle/fluid/operators/amp/update_loss_scaling_op.h]: Better `found_inf_v` -> `found_inf_data` or `found_inf_value`

**Review by zhiqiu — COMMENTED** [review]
Please add some performance results.

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cc (hunk 3 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cc]: Done.

File: python/paddle/fluid/contrib/mixed_precision/amp_nn.py (hunk 4 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/mixed_precision/amp_nn.py:73]: Thanks. This is in keeping with origin design.

File: paddle/fluid/operators/amp/update_loss_scaling_op.cc (hunk 5 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/update_loss_scaling_op.cc]: Have added.

File: paddle/fluid/operators/amp/update_loss_scaling_op.h (hunk 6 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/update_loss_scaling_op.h]: Done.

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cc [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      auto* out = outs[i];
+      out->mutable_data<T>(dev_ctx.GetPlace());
+      if (!(*found_inf_data)) {
+        framework::TensorIsfinite(*x, &is_finite);
+        *found_inf_data = !(*is_finite_data);
+      }
+      auto eigen_out = framework::EigenVector<T>::Flatten(*out);
+      auto eigen_in = framework::EigenVector<T>::Flatten(*x);
+      if (!(*found_inf_data)) {
+        eigen_out.device(dev) = eigen_in * inverse_scale;
```
Comment (gongweibao) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cc:119]: 这个性能好吗？

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu [hunk 8]
```diff
@@ -14,27 +14,32 @@ limitations under the License. */
...
+
+template <typename T>
+__global__ void CheckFiniteAndUnscale(const T* in, const T* scale, int num,
+                                      bool* found_inf, T* out) {
   const int idx = threadIdx.x + blockIdx.x * blockDim.x;
 
   if (idx < num) {
     if (!isfinite(in[idx])) {
-      *found_inf = 1;
+      *found_inf = true;
```
Comment (gongweibao) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:34]: 这个地方有问题吧？found_inf应该是全局的。多个线程之间应该共享是否找到了found_inf。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu [hunk 9]
```diff
@@ -48,18 +53,24 @@ class AmpCheckFiniteAndScaleKernel<platform::CUDADeviceContext, T>
     bool* found_inf_data = found_inf->mutable_data<bool>(dev_ctx.GetPlace());
     cudaMemset(found_inf_data, false, found_inf->numel() * sizeof(bool));
 
+    framework::Tensor inverse_scale =
+        ctx.AllocateTmpTensor<T, platform::CUDADeviceContext>({1}, dev_ctx);
+    T* inverse_scale_v = inverse_scale.template data<T>();
+
+    GpuInverse<T><<<1, 1, 0, dev_ctx.stream()>>>(scale_data, inverse_scale_v);
```
Comment (gongweibao) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:58]: 没太看懂这个inverse是干啥的？

File: python/paddle/fluid/tests/unittests/test_check_finite_and_unscale_op.py [hunk 10]
```diff
@@ -18,17 +18,17 @@
...
+        self.op_type = "check_finite_and_unscale"
         self.init_dtype()
         x = np.random.random((1024, 1024)).astype(self.dtype)
         scale = np.random.random((1)).astype(self.dtype)
 
         self.inputs = {'X': [('x0', x)], 'Scale': scale}
         self.outputs = {
             'FoundInfinite': np.array([0]),
-            'Out': [('out0', x * scale)],
+            'Out': [('out0', x / scale)],
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_check_finite_and_unscale_op.py]: 没有检查FoundInfinite的值？

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cc (hunk 7 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cc:119]: 这个是CPU的计算逻辑，AMP GPU训练不会使用到。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu (hunk 9 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:58]: 先对scale做了一次求倒数，避免在后面的循环中做除法，理论上除法开销大于乘法。

File: python/paddle/fluid/tests/unittests/test_check_finite_and_unscale_op.py (hunk 10 above)
Comment (wzzju) [reply python/paddle/fluid/tests/unittests/test_check_finite_and_unscale_op.py]: 'FoundInfinite'放在了self.outputs中，是会被检查的。

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu (hunk 8 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:34]: You are right, I will consider it. Thanks.

File: python/paddle/fluid/tests/unittests/test_update_loss_scaling_op.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        self.check_output(no_check_set=['Out'])
+
+
+class TestUpdateLossScalingOpBad(TestUpdateLossScalingOp):
+    def setUp(self):
+        self.op_type = "update_loss_scaling"
+        self.init()
+        found_inf = np.array([True], dtype=np.bool)
+        x = np.random.random((1024, 1024)).astype(self.dtype)
+        x[128][128] = np.inf
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_update_loss_scaling_op.py]: 下标改成random?

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu [hunk 12]
```diff
@@ -14,27 +14,32 @@ limitations under the License. */
...
+  *o = Inverse<T>(*s);
+}
+
+template <typename T>
+__global__ void CheckFiniteAndUnscale(const T* in, const T* scale, int num,
+                                      bool* found_inf, T* out) {
   const int idx = threadIdx.x + blockIdx.x * blockDim.x;
 
   if (idx < num) {
     if (!isfinite(in[idx])) {
```
Comment (gongweibao) [inline paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:33]: cuda支持unlikely不?

**Review by gongweibao — CHANGES_REQUESTED** [review]
另外，因为改动比较大，把Resnet50 on imagenet的性能和收敛精度都放到PR的摘要中？

File: paddle/fluid/operators/amp/check_finite_and_unscale_op.cu (hunk 12 above)
Comment (wzzju) [reply paddle/fluid/operators/amp/check_finite_and_unscale_op.cu:33]: 没有查到cuda **unlikely**的用法。

File: python/paddle/fluid/tests/unittests/test_update_loss_scaling_op.py (hunk 11 above)
Comment (wzzju) [reply python/paddle/fluid/tests/unittests/test_update_loss_scaling_op.py]: Done.

**Comment by wzzju** [comment]
> 另外，因为改动比较大，把Resnet50 on imagenet的性能和收敛精度都放到PR的摘要中？

Done.

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

File: python/paddle/fluid/contrib/mixed_precision/amp_nn.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+# limitations under the License.
+
+from paddle.fluid.data_feeder import check_variable_and_dtype, check_type
+from paddle.fluid.layer_helper import LayerHelper
+from paddle.fluid.framework import Variable
+
+__all__ = ['check_finite_and_unscale', 'update_loss_scaling']
+
+
+def check_finite_and_unscale(x, scale, name=None):
```
Comment (zhiqiu) [inline python/paddle/fluid/contrib/mixed_precision/amp_nn.py:22]: Not important.
Maybe you can change the name to `_check_finite_and_unscale` if it is not a public API in next PR.

**Review by zhiqiu — APPROVED** [review]
LGTM

File: python/paddle/fluid/contrib/mixed_precision/amp_nn.py (hunk 13 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/mixed_precision/amp_nn.py:22]: OK. I will think about it.
